Make build systems synchronize output and display commands

trws opened the pull request:
By default this will use make's `-O` flag to ensure that per-rule output is output as a chunk, and not sliced by other processes.  It also adds `--no-silent`, `V` and `VERBOSE` to make to cause commands not prefixed with `@` to be printed regardless of other options and be verbose as well as adding similar flags for ninja.  

It feels like for some cases we might actually want to go one step further and have it add `--trace` to make, so that even `@`-prefixed commands (as generated by automake in some cases) are printed and we can analyze them.  This would be step one in what @vsoch, Harshitha, @tgamblin and I have discussed about improving the build output for use with various build projects.  The next step would be looking at manipulating compiler flags to produce more complete and machinable diagnostic information from compilers, as well as making it easier to identify which executable produced the ouput (combining `--trace` with `-Oline` should make that *far* easier than it currently is).

trws committed:
set options to get make output sync and common verbose options

trws committed:
fix style

haampie commented:
By the way, `-O`/`--output-sync` is a gmake 4.0+ feature, macos is on 3.81 by default, which is 16 years old 😭. People can install gmake with spack of course...

trws commented:
Ugh... I completely forgot about that. Thanks @haampie.  Almost tempted to have spack just build it, it's not hard or long to build, but to make this as small as possible I can probably put in a couple of memoized tests, try and find a version of make that isn't broken, and if we can't then skip adding the flags.

trws committed:
test for sync support rather than blindly setting it

trws committed:
style fix

trws committed:
check separately for no-silent

haampie commented on lib/spack/spack/build_environment.py:
```diff
@@ -168,8 +169,16 @@ def __call__(self, *args, **kwargs):
                 )
                 self.supports_sync = self.returncode == 0
+                super(MakeExecutable, self).__call__(
+                    "--no-silent",
```
What's the use of `--no-silent` really? I just tried it and it doesn't overrule a `.SILENT:` target for instance.

trws replied:
Fair question, it can't override everything, but it ensures that `-s` or some of the silencing methods don't silence it.  That combined with the `V` and `VERBOSE` env vars seems to help more often than not, it's not completely required, but I don't know a better way to accomplish it.

trws committed:
switch from line to target

With jobserver passthrough support, -Otarget is a much better experience
for nested make invocations and identifying the rule that generated the
output.

trws commented:
After some poking around I decided to swap this to `-Otarget`, for single-level the `-Oline` was nice, but especially when nested `-Otarget` seemed like the right balance of getting output regularly, rather than `-Orecursive` which waits for an entire sub-make, and identifying which rule produced a given set of output, where `-Oline` produced per-rule line output, target groups the whole target together unless it's a sub-make, then each of the sub-make's targets is output as a chunk, and so-on.

@haampie, assuming this tweak and rebase passes I think this one is good to go.

trws committed:
fix style, imports and typing after rebase

trws committed:
Merge remote-tracking branch 'upstream/develop' into make-sync-and-verbose
